File: perl/print-100-to-1-shitlord.pl
# http://www.thousandtyone.com/blog/EasierThanFizzBuzzWhyCantProgrammersPrint100To1.aspx
#
# Print 100 to 1.
# That's it.
# That was the question.
# The Catch?
# You need to start with "for(int i=0;" and continue from there -
#  you cannot write anything before "for(int i=0;" and
#  you can't use two loops.

for(my $i=0;

# The "shit_lord solution" was proposed by AEvar Bjarmason,
#  (https://github.com/avar) which truly made my day.
# I thought I would adapt the idea to perl, just for fun.

    shit_lord( \$i ) > 0 ; $i-- ) {

    print $i, "\n";
}

{
    my $init;

    sub shit_lord {
        my ($i_ref) = @_;

        if ( !$init ) {
            $$i_ref = 100;
            $init   = 1;
        }
        return $$i_ref;
    }
}

File: c/print-100-to-1-shitlord.c
/*
 *  http://www.thousandtyone.com/blog/EasierThanFizzBuzzWhyCantProgrammersPrint100To1.aspx
 *
 *  Print 100 to 1.
 *  That's it.
 *  That was the question.
 *  The Catch?
 *  You need to start with "for(int i=0;" and continue from there -
 *   you cannot write anything before "for(int i=0;" and
 *   you can't use two loops.
*/

#include <stdio.h>
int main(void)
{
	for (int i = 0;
/*
 * The "shitlord solution" was proposed by AEvar Bjarmason,
 * (https://github.com/avar) which truly made my day.
 * I thought I would actually code it up in C, just for fun.
 */
	     shitlord(&i); --i) {
		printf("%d\n", i);
	}
	return 0;
}

static int shitlord_init = 0;
int shitlord(int *i)
{
	if (!shitlord_init) {
		*i = 100;
		shitlord_init = 1;
	}
	return *i;
}
